safehttp/mux.go

```diff
@@ -71,20 +71,42 @@ type ServeMux struct {
 	mux *http.ServeMux
 }
 
-func registerHandlers(mux *http.ServeMux, handlers map[string]map[string]handlerConfig) {
-	for pattern, handlersPerMethod := range handlers {
-		handlersPerMethod := handlersPerMethod
-		mux.HandleFunc(pattern, func(w http.ResponseWriter, r *http.Request) {
-			cfg, ok := handlersPerMethod[r.Method]
-			if !ok {
-				http.Error(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
-				return
-			}
-			processRequest(cfg, w, r)
-		})
+func registerHandlers(mux *http.ServeMux, handlers map[string]map[string]*handlerConfig) {
+	for pattern, methodToHandlerConfigMap := range handlers {
+		httpHandler := generateHttpHandlerFunc(methodToHandlerConfigMap)
+		mux.HandleFunc(pattern, httpHandler)
+	}
+}
+
+// generateHttpHandlerFunc converts a handlerConfig to a classic http.HandlerFunc.
+// We need to have this map because http.Handle handles requests for any HTTP request method per default,
+// which we want to avoid in safehttp.
+func generateHttpHandlerFunc(methodToHandlerConfig map[string]*handlerConfig) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		cfg, ok := methodToHandlerConfig[r.Method]
+		if !ok {
+			http.Error(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
+			return
+		}
+		processRequest(cfg, w, r)
 	}
 }
 
+// HttpHandlerForTransition generates a classic http.Handler using its ServeMuxConfig
+// (i.e. already installed Interceptors etc).
+// This is intended to be used during transition from using the http package to safehttp.
```

Comment:
How are you going to support the case of two handlers registered for the same path but different methods?

Reply:
Very good catch... I'll think about it.

Reply:
Seems like we will need to expose the private stdHandler here and and make users pass a method -> handler map or so.
Or live with the fact that we are doing things different from `http` and ask users of the transition API to write method-aware handlers (like they have to do in standard http) -- this seems very bad, as long as we want to stick with a per-method handler registration API in safehttp.

Reply:
PTAL, I pushed a new version that should work (in an arguably ugly way). I'm not convinced yet that letting users handle this (like in example above) isn't the best way.